### ISSUE TITLE
Unexpected variable state during column insertion

### ISSUE THREAD

**Issue by ysu1028** [atoptima/Coluna.jl#654]
Hello,

I'm trying to initialize coluna with multiple initial columns but I got the error "Unexpected variable state during column insertion" when I input more than 2 columns (i.e., more than one entire solutions). The code of my callback for the initialization is:

```julia
function callback_for_initial_cols(cbdata)
    #### columns: the inital columns for my personal generator ###########
    for i in 1:length(columns)
        #### for each column in columns, compute the cost of column, generate non-zero variables and the values of variables ####
        MOI.submit(
            model,
            BlockDecomposition.PricingSolution(cbdata),
            cost,  # cost of column
            non_zero_var, # non-zero variables
            var_value  # values of variables
        )
    end
end
```

In the above callback, I submit columns one by one via MOI.submit.

**Question 1:** Can Coluna accept multiple columns for initialization?

I also met another problem during the initialization process, it seems if I submit columns one by one, Coluna will not start with mlp = my initial solution cost. Instead, it still starts with a big mlp value (I don't know how it gets this value) and I'm not sure if my columns are really submitted or not (also it starts with an infeasible solution missing one customer node, but I input the feasible solution, see in the first Figure below). However, if I concatenate the two columns of my solution into one column (now MOI.submit outside the loop and non_zero_var, var_value record the concatenated column), and then use MOI.submit to submit the "big" column, it can start with mlp = my initial solution cost. 

I attached the Figure:
For the first way of initialization:

<img width="1152" alt="截屏2022-04-12 10 16 13" src="https://user-images.githubusercontent.com/58483379/162914093-d71e5a2e-6a4f-4623-8f1e-82aaf2188363.png">

For the second way of initialization:

<img width="1160" alt="截屏2022-04-12 10 12 48" src="https://user-images.githubusercontent.com/58483379/162913576-eb4d6518-12e9-4a56-94da-96d353bd890b.png">

**Question 2:** Is the second way equal to the first way? I feel a little strange because actually in the second way, the concatenated solution is not a feasible solution. Instead, the first way initializes with feasible columns, normally it should take these columns and start with a good MPL value...

**Question 3:** From the second question, I'm wondering if no feasibility check for inputted columns is conducted in Coluna?

My version is:
Coluna v0.3.12
JuMP v0.21.10
Gurobi v0.9.14
BlockDecomposition v1.4.0

Thank you very much for you time and looking forward to hearing from you!

Best reagrds,
Yue

**Comment by ysu1028** [comment]
@guimarqu

**Comment by ndupin** [comment]
Hi,

Migrating to version 0.4 of Coluna, there is an error now for initialization/heuristic callbacks:

ERROR: LoadError: Coluna.Algorithm.MissingPricingDualBound()

How can we give a (trivial) dual bound or indicate that the dual bound will not be updated?

**Comment by guimarqu** [comment]
Hi,  

when you do heuristic pricing/column initialization using the pricing callbacl and you don't know the contribution of the subproblem to the lagrangian bound, you can submit `-Inf`: 

```julia
 MOI.submit(model, BD.PricingDualBound(cbdata), -Inf)
```

When the pricing is exact, you should pass the cost of the column (single column pricing) or the cost of the best column found (multiple column pricing).

**Comment by ysu1028** [comment]
> 

Hello,

Thank you very much for the very detailed reply!

I've updated my Coluna to version 0.4.0 as you suggested, however, the "Unexpected variable state during column insertion" error still exists if I submit more than one feasible solutions (each solution with two columns).

I attached the screenshot of the results:

1. I submit sequentially the columns (I make sure not to submit the identical column) and I submit one initial solution

<img width="908" alt="截屏2022-04-13 10 36 55" src="https://user-images.githubusercontent.com/58483379/163135802-5a53569e-fc0c-44ac-8f5b-edaba7d196e9.png">

2. the final results of coluna with initialization is:
<img width="893" alt="截屏2022-04-13 10 39 05" src="https://user-images.githubusercontent.com/58483379/163136076-39a61189-766e-422d-bd99-b4d227236939.png">

The final result of Coluna + initialization is 10237.38

However, if I run Coluna directly and do not use the initialization callback. Just do:

coluna = optimizer_with_attributes(
    Coluna.Optimizer,
    "params" => Coluna.Params(
        solver = Coluna.Algorithm.TreeSearchAlgorithm() # default BCP
    ),
    "default_optimizer" => Gurobi.Optimizer # Gurobi for the master & the subproblems
)

and 

master = getmaster(decomposition)
subproblems = getsubproblems(decomposition)
specify!.(subproblems, lower_multiplicity = 0, upper_multiplicity = numb_veh)
optimize!(model)

Then, I get different results:
<img width="948" alt="截屏2022-04-13 10 45 31" src="https://user-images.githubusercontent.com/58483379/163137343-d11d11a7-aec8-419c-b3a5-48b6783a29fb.png">

in this case, coluna solves into optimality.

And the two results have a difference of 10000.

Based on these observations:
1. I guess the "unexpected insertion" error is not on if I submit identical columns or not
2. I don't understand why the two results are different and just have a difference of 10000. I'm very curious about it, do you have any hints for that?

Thank you again for your time and looking forward to hearing from you!!

Best regards,
Yue

**Comment by ndupin** [comment]
Yue, it will be easier for me to look at it as I have the code, I'll do it this weekend.

**Comment by guimarqu** [comment]
The "Unexpected variable state during column insertion" happens when you try to insert a column that already exists in the master and that is already active. It usually means that there is a bug in the pricing callback. Indeed, you submit a column with negative reduced cost that is already in the master. However, an active master column cannot have a negative reduced cost.

In your first run, it turns out your instance is infeasible (the difference of 10000 is the cost of an artificial variable).
In the second, it seems like ok.

Do you use the same model in both cases? If yes, I am quite surprised. Can a bad initialization lead to infeasibility? I would have said no but it requires further investigation.

**Comment by ndupin** [comment]
I may have an explanation of the origin of the bug. 
In this initialization callback, some non-null variables were not considered in the MOI.submit. It was forgotten as these variables have no cost in the objective function (necessary just for feasibility, in a rich model). The reduced cost of these variables may be non null, and thus we would have to consider it in the MOI.submit. Do you confirm it is the good practice?  

Do you confirm it may be the source of the error?

It raises other questions, on the best way to use Coluna in this case. We have a labelling algorithm for the exact sub-problem, so we do not need to have the exact MILP model for using the default callback with the MILP solver. As in the CVRP demo, another modeling of the master problem would  help to have more compact columns (without useless information for feasibility) and more easier callbacks to write (without considering the additional variables for feasibility of subproblems), considering the rich feasibility constraints only in the callback. It may also have an influence for a branch-and-cut-and-price extension. 

I see then 2/3 possibilities to write the master problem and the related callbacks for this rich problem, I would need for advise of good practice using Coluna to choose among these 2/3 possibilities. Maybe some "tutorials" on the good/best practices to use Coluna would be useful.

**Comment by guimarqu** [comment]
Yes, you have to consider all subproblems variables. If you forgot some non-zero variables in your columns, it's a bug. I think the correct columns should have been generated but I don't know what can happen if you add infeasible columns.
So it may be the origin of the bug.

For the second question, it really depends on what you want to do. I think it's pretty complicated to say what is a good practice. I would say to design your application in order to extend it easily and experiment with several models.

**Comment by ysu1028** [comment]
> Yes, you have to consider all subproblems variables. If you forgot some non-zero variables in your columns, it's a bug. I think the correct columns should have been generated but I don't know what can happen if you add infeasible columns. So it may be the origin of the bug.
> 
> For the second question, it really depends on what you want to do. I think it's pretty complicated to say what is a good practice. I would say to design your application in order to extend it easily and experiment with several models.

Thank you for the reply!

In fact, we have several types of decision variables in the compact formulation (binary, continuous ...). In the objective function, only two kinds of decision variables are related (one is x_i,j^k, other is R_i^k), other decision variables appear in the constraints (the values of these decision variables are used to check feasibility). For the initialization, as you mentioned, no feasibility check is embedded in Coluna, it would be enough to just feed Coluna with objective-related non-zero variables (i.e., x[i,j] and R_i ). Thus, I guess the problem may not on we didn't submit the values for other feasibility-related decision variables.

By the way, I'm reading the updated document of initial column (https://atoptima.github.io/Coluna.jl/dev/start/initial_columns/) and I'm implementing it on my computer. It seems the BlockDecomposition.InitialColumn(cbdata) and BlockDecomposition.InitialColumnsCallback() are not activated yet. Could you tell me when these functions will be activated? It's very nice to see the updated document and new functions!

Thank you very much and have an excellent afternoon!

Best regards,
Yue

**Comment by ysu1028** [comment]
Dear all,

It seems adding **upper_multiplicity = numb_veh** in the argument BlockDecomposition.specify! solve the problem. Now the coluna model with initialization callback converges to the same value of only using coluna. Also, multiple initial columns can be added. No more "unexpected variable" error.

Best,
Yue

**Comment by ysu1028** [comment]
Dear Guillaume,

First, thank you very much for your help with my questions!

Here is a further question from me :-P

I found that the number of columns "cols" printed in the log is sometimes less than that I generated from my solver. I guess there are some functions to filter the columns I submit. 

 If it is true:
=> Could you tell me where are these functions in the codes?
=> Would these functions be time-consuming when I submit a large number of columns (>1000 columns)?

Thank you very much and looking forward to hearing from you!

Best regards,
Yue

**Comment by ysu1028** [comment]
@guimarqu

**Comment by ysu1028** [comment]
Dear Guillaume,

Thank you very much for your reply!

I'll send you the Profile results by email. The format is in *.jlprof, you can visualize it by adding package ProfileView.jl and typing "ProfileView.view(nothing)" to open a blank window which you can populate with saved data by clicking on the "open" icon.

I hope that would be helpful!

P.s. "mycolgen" is my pricing algorithm and "BP" is also my algorithm;

Best,
Yue

**Comment by guimarqu** [comment]
How many columns do you insert? According to your profiling, it seems to be not so bad but it depends a lot on the number of columns added. 

You didn't use the initial column callback, did you?

![Capture d’écran 2022-05-02 à 18 31 49](https://user-images.githubusercontent.com/5951111/166282420-a777a55c-27a4-4ecb-b4f3-75a83968d666.png)

**Comment by guimarqu** [comment]
@artalvpes significantly improved the solution pool implementation. It will be merged today (#660).
